web/modules/custom/brms_sor/src/Service/FmsSorApi.php

```diff
@@ -0,0 +1,252 @@
+   * @var string
+   */
+  protected $baseUrl;
+
+  /**
+   * FMS API client id.
+   *
+   * @var string
+   */
+  protected $clientId;
+
+  /**
+   * Secret of the DnB client.
+   *
+   * @var string
+   */
+  protected $clientSecret;
+
+  /**
+   * Constructs a FmsSorApi object.
+   *
+   * @param \GuzzleHttp\ClientInterface $http_client
+   *   The GuzzleHttp Client.
+   * @param \Drupal\Core\Logger\LoggerChannelFactoryInterface $logger_factory
+   *   The logger factory.
+   * @param \Drupal\Core\Cache\CacheBackendInterface $cache_backend
+   *   The cache backend.
+   * @param \Drupal\key\KeyRepositoryInterface $key_repository
+   *   The key repository.
+   * @param \Drupal\Component\Serialization\Json $json_service
+   *   The JSON serializer service.
+   * @param \Drupal\Component\Datetime\TimeInterface $time
+   *   The time service.
+   */
+  public function __construct(
+    protected ClientInterface $http_client,
+    protected LoggerChannelFactoryInterface $logger_factory,
+    protected CacheBackendInterface $cache_backend,
+    protected KeyRepositoryInterface $key_repository,
+    protected Json $json_service,
+    protected TimeInterface $time,
+  ) {
+    $this->httpClient = $http_client;
+    $this->logger = $logger_factory;
+    $this->cache = $cache_backend;
+    $this->json = $json_service;
+    $settings = Settings::get('brms_sor');
+    $this->authUrl = $settings['fms_api_auth_url'];
+    $this->baseUrl = $settings['fms_api_base_url'];
+    $this->clientId = $settings['fms_api_client_id'];
+    $this->clientSecret = $key_repository->getKey('fms_api')->getKeyValue();
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getAccessToken(): ?string {
+    $cache_access_token = $this->cache->get('brms_sor_fms_access_token');
+    // If there is a cached token, and it's not going to expire in the next
+    // minute, return it, otherwise generate a new one.
+    if ($cache_access_token && $cache_access_token->expire > ($this->time->getRequestTime() + 60)) {
+      return $cache_access_token->data;
+    }
+
+    try {
+      $response = $this->httpClient->request('POST', $this->authUrl, [
+        'headers' => [
+          'Content-Type' => 'application/x-www-form-urlencoded',
+        ],
+        'form_params' => [
+          'grant_type' => 'client_credentials',
+          'client_id' => $this->clientId,
+          'client_secret' => $this->clientSecret,
+        ],
+      ]);
+
+      $data = $this->json->decode($response->getBody()->getContents());
+      $this->cache->set('brms_sor_fms_access_token', $data['access_token'], $this->time->getRequestTime() + $data['expires_in']);
+
+      return $data['access_token'];
+    }
+    catch (\Exception $e) {
+      $this->logger->get('brms_sor')->error('Failed to get access token from the FMS API: @message', ['@message' => $e->getMessage()]);
+      return NULL;
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function get(string $call, array $data = []): array {
```

Comment:
isn't get and patch basically the same function? you could do request with $mode 'GET' or 'PATCH'

Reply:
I merged get and patch into a single apiCall method